### ISSUE TITLE
Change default "Redo" to `cs-z`

### ISSUE THREAD

**Issue by kendonB** [dictation-toolbox/Caster#426]
AFAIK, more text editors on Windows use `cs-z` and not `c-y` for "Redo". Though I may be wrong. However, "Redo" should be included in each App, if the app redo differs from the default in `nav.py`.

So, this issue is about:
 - deciding whether to use `cs-z` or `c-y` as the default
 - including the other one in all apps for which the default does not work

**Comment by LexiconCode** [comment]
That's Reasonable.
Why don't we take the time to test each app. To see if `cs-z` is the preferred method  "Redo". As you test post name of the apps and I'll check them off here.

List of Apps.
- [ ] atom
- [ ] chrome
- [ ] eclipse
- [ ] emacs
- [ ] excel
- [ ] explorer
- [ ] firefox
- [ ] flashdevelop
- [ ] foxitreader
- [ ] gitter
- [ ] jetbrains
- [ ] kdiff3
- [ ] lyx
- [ ] msvc
- [ ] notepadplusplus
- [ ] outlook
- [ ] rstudio
- [ ] sqldeveloper
- [ ] ssms
- [ ] sublime
- [ ] totalcmd
- [ ] typora
- [ ] visualstudio
- [ ] vscode
- [ ] winword

**Comment by kendonB** [comment]
Some apps allow for both, so there should be two lists.

```
rstudio:         cs-z works and c-y fails
chrome:          cs-z works and c-y works
firefox:         cs-z works and c-y works
notepadplusplus: cs-z works and c-y works
outlook:         cs-z fails and c-y works
winword:        cs-z fails and c-y works
excel:           cs-z fails and c-y works
explorer:        cs-z fails and c-y works
```

**Comment by Danesprite** [comment]
There is an [aenea](https://github.com/dictation-toolbox/aenea) class I use in these situations called [ContextAction](https://github.com/dictation-toolbox/aenea/blob/ef8ef3d51df1e58aba6eec3ef5e03400646625de/client/aenea/wrappers.py#L273). It executes different actions depending on the first dragonfly context in a list that matches, falling back on a default action. It isn't dependent on any part of aenea to work. 

An example for "redo" would be:
```Python
redo = ContextAction(default=Key("c-y"), actions=[
    # Use cs-z for rstudio
    (AppContext(executable="rstudio"), Key("cs-z")),
])

```

I find it very useful as an Emacs user. Maybe caster could use something like this?

**Comment by LexiconCode** [comment]
> There is an [aenea](https://github.com/dictation-toolbox/aenea) class I use in these situations called [ContextAction](https://github.com/dictation-toolbox/aenea/blob/ef8ef3d51df1e58aba6eec3ef5e03400646625de/client/aenea/wrappers.py#L273). It executes different actions depending on the first dragonfly context in a list that matches, falling back on a default action. It isn't dependent on any part of aenea to work.

I love the idea and it's something that's really needed. There are other inconsistencies besides Redo between editors that could be fixed by such an implementation.

However it will need to be reworked a bit. ensure_execution_context needs to be eliminated and context needs to utilize AppContext

**Comment by Danesprite** [comment]
Oh oops. I didn't notice that function was used there. It should be possible to use `Window.get_foreground()` instead if there is no data.

Yep it simplifies things a lot, especially with CCR.

**Comment by LexiconCode** [comment]
After talking to @Danesprite he is looking into including ContextAction as part of the Dragonfly framework.
Even with ContextAction implemented it is still worthwhile changing windows use cs-z for "Redo". The exceptions can be handled with ContextAction.

**Comment by LexiconCode** [comment]
[ContextAction ](https://dragonfly2.readthedocs.io/en/latest/actions.html?%20#dragonfly.actions.action_context.ContextAction) class is now part of dragonfly v13. Academics are going to keep me busy for the next few weeks but I wanted let you know I hadn't forgotten about this.

**Comment by kendonB** [comment]
#493 starts to fix this issue. However, we need to other add or check other apps. i.e. Emacs. Emacs will be a little tricky, however, because Emacs users can choose between "traditional" and "Emacsy" undo and redo commands, I believe. Someone should correct me if I'm wrong. And we should talk about a way to handle that particular user preference e.g. using an entry in `settings.toml`

**Comment by LexiconCode** [comment]
I'm not familiar with using Emacs. Is this part of the setup process to choose between "traditional" and "Emacsy"or is there a default? If it's part of the install process something like ` settings.toml` will have to be used to specify.

**Comment by kendonB** [comment]
Emacsy is the default in the distributions I've tried. The alternative is known as Cua mode: https://www.emacswiki.org/emacs/CuaMode

I also shouldn't call it "traditional" since Emacs is a very old program and probably predates the Windowsy options.

**Comment by LexiconCode** [comment]
As a general rule of thumb it's best to support application defaults in this case Emacsy.

**Comment by LexiconCode** [comment]
What's the current status of this issue?

**Comment by kendonB** [comment]
This can be closed once we add Emacs to the action. There is the bigger design question of whether to use `ContextAction` for these types of things or redefine them in the particular app files as in https://github.com/dictation-toolbox/Caster/pull/530. I actually think I prefer the latter approach.

**Comment by LexiconCode** [comment]
To clarify, defining them in particular app files?

**Comment by kendonB** [comment]
Yes just like this one is "redefined" in their right now (obviously this particular one was redefined in error as it's actually the same in firefox). 

https://github.com/dictation-toolbox/Caster/blob/36345de83f49010db5c402e4c0768e42b845a5bf/castervoice/apps/browser/browser_shared_commands.py#L50-L51

https://github.com/dictation-toolbox/Caster/blob/36345de83f49010db5c402e4c0768e42b845a5bf/castervoice/apps/browser/firefox.py#L13-L14

**Comment by LexiconCode** [comment]
The grammar standard API have been removed except as an example. The reasoning for for this change can be found [here](https://github.com/dictation-toolbox/Caster/issues/533#issuecomment-560036137).

Therefore we can utilize `ContextAction` or defined them individually within applications.
